### PR TITLE
Add `Tooltip` support to `Button`

### PR DESCRIPTION
**Pull request by hectahertz**
### What are you trying to accomplish?

The API of `Button` should be updated so `Tooltip` can be rendered via the component arguments.

- Closes https://github.com/github/primer/issues/776
- See the Accessibility story issue https://github.com/github/accessibility/issues/802

### What approach did you choose and why?

We decided to use a slot to render the `Tooltip` as it allows for explicit control of the defaults, which we want to ensure proper integration and an accessible outcome. We will follow the same pattern for the rest of the components that integrate `Tooltip`

For example, we set `type: :description` as the default, as most buttons will have a label already.

### Docs
Screen Shot 2022-03-15 at 13 03 10

Screen Shot 2022-03-17 at 14 33 00

### CODE REVIEW

File: app/components/primer/button_component.rb [hunk 1]
```diff
@@ -56,6 +56,21 @@ class ButtonComponent < Primer::Component
     }
     alias counter trailing_visual_counter # remove alias when all buttons are migrated to new slot names
 
+    # `Tooltip` that appears on mouse hover or keyboard focus over the button. Use tooltips sparingly and as a last resort.
+    # Consult the <%= link_to_component(Primer::Alpha::Tooltip) %> documentation for more information.
+    #
+    # @param system_arguments [Hash] Same arguments as <%= link_to_component(Primer::Alpha::Tooltip) %>.
```
Comment (khiga8) [inline app/components/primer/button_component.rb:62]:
```suggestion
    # @param type [Symbol] (:description) <%= one_of(Primer::Alpha::Tooltip::TYPE_OPTIONS) %>
    # @param system_arguments [Hash] Same arguments as <%= link_to_component(Primer::Alpha::Tooltip) %>.
```

This adds documentation for the default `type`

File: app/components/primer/button_component.rb [hunk 2]
```diff
@@ -56,6 +56,21 @@ class ButtonComponent < Primer::Component
...
+    renders_one :tooltip, lambda { |**system_arguments|
```
Comment (khiga8) [inline app/components/primer/button_component.rb:63]:
```suggestion
    renders_one :tooltip, lambda { |text:, **system_arguments|
```

We may want to explicitly surface the required `text:` argument here. If we do this, we'll want to add `@param` to document this.

File: app/components/primer/button_component.rb [hunk 3]
```diff
@@ -56,6 +56,21 @@ class ButtonComponent < Primer::Component
     }
     alias counter trailing_visual_counter # remove alias when all buttons are migrated to new slot names
 
+    # `Tooltip` that appears on mouse hover or keyboard focus over the button. Use tooltips sparingly and as a last resort.
+    # Consult the <%= link_to_component(Primer::Alpha::Tooltip) %> documentation for more information.
```
Comment (khiga8) [inline app/components/primer/button_component.rb:60]: I think we may want to call extra ttention to the `type:` here. Maybe something like:

```suggestion
    # **Important:** This tooltip defaults to `type: :description`. In a few scenarios, `type: :label` may be more appropriate.
    # Consult the <%= link_to_component(Primer::Alpha::Tooltip) %> documentation for more information.
```

File: app/components/primer/button_component.rb [hunk 4]
```diff
@@ -96,6 +111,12 @@ class ButtonComponent < Primer::Component
     #     Button
     #   <% end %>
     #
+    # @example With tooltip
+    #   <%= render(Primer::ButtonComponent.new(id: "button-with-tooltip")) do |c| %>
+    #     <% c.tooltip(text: "Tooltip text") %>
+    #     Button
+    #   <% end %>
```
Comment (khiga8) [inline app/components/primer/button_component.rb:118]: Maybe this is overkill, but should we link to `Tooltip` documentation here too using `@code` and `@description`?

**Review by khiga8 — COMMENTED** [review]
I added a few comments.
Overall this is looking good!!

File: app/components/primer/button_component.rb (hunk 4 above)
Comment (hectahertz) [reply app/components/primer/button_component.rb:120]: Not overkill at all, good idea!

File: app/components/primer/button_component.rb (hunk 2 above)
Comment (hectahertz) [reply app/components/primer/button_component.rb:65]: I thought about this, but then I saw `type` and `for_id` are required too.

I decided keeping the definition of which parameters are required on `Tooltip` will be easier to maintain. I'll leave it as is for now if that's ok.

Comment (khiga8) [reply app/components/primer/button_component.rb:65]: That makes sense 👍

**Review by khiga8 — APPROVED** [review]
Nice work 💪

**Review by jonrohan — APPROVED** [review]

**Pull request merged**